tpm/src/emulator.rs

```diff
@@ -0,0 +1,450 @@
+const PTM_CAP_RESET_TPMESTABLISHED: u64 = 1 << 7;
+const PTM_CAP_STOP: u64 = 1 << 10;
+const PTM_CAP_SET_DATAFD: u64 = 1 << 12;
+const PTM_CAP_SET_BUFFERSIZE: u64 = 1 << 13;
+
+///Check if the input command is selftest
+///
+pub fn is_selftest(input: Vec<u8>, in_len: usize) -> bool {
+    if in_len >= TPM_REQ_HDR_SIZE {
+        let ordinal: &[u8; 4] = input[6..6 + 4]
+            .try_into()
+            .expect("slice with incorrect length");
+
+        return u32::from_ne_bytes(*ordinal).to_be() == 0x143;
+    }
+    false
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Input socket path for tpm emulator does not exist")]
+    SocketPathExists(),
+    #[error("Could not initialize emulator's backend")]
+    InitializeEmulator(#[source] anyhow::Error),
+    #[error("Failed to create data fd to pass to swtpm")]
+    PrepareDataFd(#[source] anyhow::Error),
+    #[error("Failed to run Ctrl Cmd: {0}")]
+    RunCtrlCmd(#[source] anyhow::Error),
+    #[error("Emulator doesn't implement min required capabilities: {0}")]
+    CheckCaps(#[source] anyhow::Error),
+    #[error("Emulator failed to deliver request: {0}")]
+    DeliverRequest(#[source] anyhow::Error),
+    #[error("Emulator failed to send/receive msg on data fd: {0}")]
+    SendReceive(#[source] anyhow::Error),
+}
+
+type Result<T> = anyhow::Result<T, Error>;
+
+#[derive(Clone)]
+pub struct BackendCmd {
+    pub locty: u8,
+    pub input: Vec<u8>,
+    pub input_len: usize,
+    pub output: Vec<u8>,
+    pub output_len: usize,
+    pub selftest_done: bool,
+}
+
+pub struct Emulator {
+    had_startup_error: bool,
+    cmd: Option<BackendCmd>,
+    caps: PtmCap, /* capabilities of the TPM */
+    ctrl_socet: SocketDev,
+    data_fd: RawFd,
+    established_flag_cached: bool,
+    established_flag: bool,
+}
+
+impl Emulator {
+    /// Create Emulator Instance
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - A path to the Unix Domain Socket swtpm is listening on
+    ///
+    pub fn new(path: String) -> Result<Self> {
+        if !Path::new(&path).exists() {
+            return Err(Error::InitializeEmulator(anyhow!(
+                "The input TPM Socket path: {:?} does not exist",
+                path
+            )));
+        }
+        let mut socket = SocketDev::new();
+        socket.init(path).map_err(|e| {
+            Error::InitializeEmulator(anyhow!("Failed while initializing tpm emulator: {:?}", e))
+        })?;
+
+        let mut emulator = Self {
+            had_startup_error: false,
+            cmd: None,
+            caps: 0,
+            ctrl_socet: socket,
+            data_fd: -1,
+            established_flag_cached: false,
+            established_flag: false,
+        };
+
+        emulator.prepare_data_fd()?;
+
+        emulator.probe_caps()?;
+        if !emulator.check_caps() {
+            emulator.had_startup_error = true;
+        }
+
+        if !emulator.get_established_flag() {
+            emulator.had_startup_error = true;
+        }
```

Comment:
It would be better to just return an `Err(..)` here if there was failure.

Reply:
makes sense. Updated.